### ISSUE TITLE
Wildcarding assets in pubspec.yaml (or allow an entire directory to be included)

### ISSUE THREAD

**Issue by apwilson** [flutter/flutter#4890]
I'd like to be able to add an entire folder of assets without having to enumerate each one in my flutter.yaml.

**Comment by eseidelGoogle** [comment]
Here is the handling of the `assets` key, probably pretty easy to add this (and make other improvements to this logic):
https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/asset.dart#L343

**Comment by eseidelGoogle** [comment]
We're looking to remove .flx files entirely in https://github.com/flutter/flutter/issues/9900 which will likely change how we specify assets.

**Comment by dragonfax** [comment]
Still no solution for this? I have a lot of images in my app, and I'm adding more all the time.

**Comment by leedstyh** [comment]
Before fixing, we can write a script use your favorite script language (say dart is good choice) to automatically generate the `.yaml` file.

**Comment by sethladd** [comment]
cc @mit-mit for dev experience

**Comment by sannysoft** [comment]
Would be great if we can either add wildcard paths/directories

**Comment by MichaelSowah** [comment]
is there any progress on this... been more than a year now still nothing or has it been put on hold

**Comment by Hixie** [comment]
We've removed FLXes but we haven't yet implemented wildcards. I'll see if we can increase its priority.

We would naturally be happy to review patches if any of you are interested in implementing this before we get to it. :-)

**Comment by MichaelSowah** [comment]
@Hixie has the FLXes removals landed on the dev channels yet?

**Comment by floriangbh** [comment]
Really urgent feature, imagine when you have a folder of static web page with more than 3000 files.. :(

**Comment by zoechi** [comment]
@floitschG you can use code generation to add them to the `pubspec.yaml` file.
Btw: I think they should not be in `pubspec.yaml`

**Comment by floriangbh** [comment]
@zoechi Where you think they whould be put ?

**Comment by zoechi** [comment]
@floriangbh currently they need to be in `pubspec.yaml` 
I just think it wasn't a good idea to require them to be in this file.

**Comment by sethladd** [comment]
Someone please create a simple Dart script that scans a directory, finds all the assets, and adds them to a pubspec.yaml :) You can use this library to parse the YAML file: https://pub.dartlang.org/packages/yaml

The community will really thank you :)

**Comment by floriangbh** [comment]
That's not a Dart script, but a simple sh command : 
`find ./assets -name "*.*" -print | grep '\.' | sed 's/.\/assets/-\ assets/g' >> assets.yaml`  

In fact , it not resolve my problem, if there are too much file we can't run project : https://github.com/flutter/flutter/issues/15486.. :(

**Comment by fmatosqg** [comment]
@sethladd I have a POC for such a script, but it's instead scanning pubspec and generating a dart file with static fields, one for each file, to avoid runtime errors. Big static language fan here. 

I would love to have it the other way around, 
a) no mention at all in pubspec
b) by convention pick folder named `assets` or whatever other name
c) scan folder for images, json, fonts, etc
d) integrate with whoever used to read that section from pubspec (AssetBundle?)
e) generate dart file with static class / fields for avoiding typing strings in application code

This is probably out of scope for this issue, but I'd like to work on it. I'm mostly interested in the last step, and to make it happen seamlessly I'd need to tap into flutter internal code since there's no way to publish a package and run it as part of a pre-build task.

**Comment by fmatosqg** [comment]
Good news is that I have a working version,
- new pubspec parameter ```enable-asset-scan: true``` will add to bundle every file found under assets/images/* and because of variant support that extends also to assets/images/*/* 
- new pubspec parameter modifies default scan path

I guess that it needs some polishing, for example I made it thinking about images only, but there's also support for other types of assets. So maybe the pubspec parameter for path should be an array or something more complex?

BTW, fonts are slightly more complex with Bold and Light weights. Code is easy to extend but I'm not sure how to write the specs.

Bad news is that I'll need a few days to submit PR due to non-technical reasons.

**Comment by Hixie** [comment]
cc @tvolkert

**Comment by tvolkert** [comment]
@szakarias is this the issue you mentioned regarding allowing entire directories of assets to be included in the pubspec, or is there a dup?

**Comment by szakarias** [comment]
This is it.

**Comment by Cyan101** [comment]
This really should get done soon

**Comment by MaskyS** [comment]
Hope to see this get done soon!

**Comment by MaskyS** [comment]
@Hixie , how much work is left for this after #16413 ?

**Comment by tvolkert** [comment]
This should be done now.  You can list a folder of assets by having it include a trailing slash.

**Comment by dragonfax** [comment]
awesome

**Comment by MaskyS** [comment]
@tvolkert Wilcarding assets is still not possible though? Will that be implemented in a follow-up issue?

**Comment by fmatosqg** [comment]
No wildcards or subdirectories yet. There are a couple of wrinkles before having them done properly, I plan to be verbose about it on a medium post soon.

Follow ups on that direction are likely on my part but my time is very short, and I'm not part of flutter team so naturally I can't guarantee my ideas will be accepted. Be patient, time frames here were weeks but can get even longer.

Eventually getting the dialog started and showing there's interest will help.

**Comment by fmatosqg** [comment]
Details on why it does what it does, and why it doesn't what it doesn't (yet). Also there's a shortlist of things related to this work.
 https://medium.com/@fmatosqg/contributing-to-flutter-part-3-4b96d924d8ba

One more time, thanks @tvolkert for helping me all along.

cc @MaskyS

**Comment by bunopus** [comment]
According to documentation, now it's possible to add entire _assets_ directory:

> To include all assets under a directory, specify the directory name with the / character at the end:
```
flutter:
  assets:
    - assets/
```
but it gives analyser error:
`warning: The asset assets/ does not exist. (asset_does_not_exist at [my_project] pubspec.yaml:31)`

**Comment by zoechi** [comment]
@bunopus seems to be only in `master` yet. See #16413
You also need flutter/flutter-intellij#2384 which will probably included in the next Flutter IntelliJ plugin update. 

Not sure about VSCode @DanTup do you know about that?

**Comment by bunopus** [comment]
@zoechi thanks! Found related bug https://github.com/dart-lang/sdk/issues/33550
Ignore it for now, but it's really confusing, because it's in the docs on https://flutter.io/assets-and-images/

**Comment by DanTup** [comment]
Looks like the fix is in the analysis server, so VS Code doesn't require any changes. You'll get the fix when you get a new Flutter that has a new SDK in it (same as for IntelliJ) 😄

**Comment by fmatosqg** [comment]
@DanTup hey I was going through https://github.com/dart-lang/sdk/issues/33550 but that's a workaround. Are you aware of a definitive fix for this? if not, I'd like to open new ticket/reopen the above one, and in the mean time add the workaround above in the documentation as a PR.

**Comment by DanTup** [comment]
@fmatosqg The issue is fixed (https://github.com/dart-lang/sdk/commit/d895b4913420036b24110376367946f6195f478b) - the workaround is a temporary fix you can use until you get a version of Flutter that includes the fix.

**Comment by fmatosqg** [comment]
@DanTup awesome :)
Is there a easy way to know when that change trickles down to flutter master, without checking dart commit hash with `flutter --version` and cloning dart to track it? I understand the idea but not sure if there's an easy way to know, and occasionally people come to me asking about this issue.

I understand that is not usually needed though.

Bot idea: register your email for when a given dart/flutter commit reaches master / dev / beta

**Comment by DanTup** [comment]
@fmatosqg If you click on the changeset above, you can see the tags shown on GitHub:

![screen shot 2018-07-10 at 4 30 44 pm](https://user-images.githubusercontent.com/1078012/42520643-a105e1fa-845e-11e8-9bc2-1886bd27cbc0.png)

The oldest one there is the one you care about (dev64 in this case).

When you run `flutter --version`, it includes the Dart SDK version:

```text
dantup-macbookpro:Dart-Code dantup$ flutter --version
Flutter 0.5.7-pre.83 • channel master • https://github.com/DanTup/flutter.git
Framework • revision b931640c1d (10 hours ago) • 2018-07-10 06:51:12 +0100
Engine • revision 6fe748490d
Tools • Dart 2.0.0-dev.63.0.flutter-4c9689c1d2 <------
```

You can see this one says `dev.63`. So it's not in master yet, but you should be able to tell when it is fairly easily.

> Bot idea: register your email for when a given dart/flutter commit reaches master / dev / beta

Yeah, a generic GitHub bot you can mention on an issue that pings you when it hits a release would be awesome; but hard to implement (hard to generically tie issues to releases). A more specific one (eg. just Flutter) would be easier, but less awesome!

**Comment by jayrmotta** [comment]
Guys, the only issue with this new implementation is that it only gets the repository mentioned, it would be good to be able to do this recursively since I'm trying to package a web application there during the migration to flutter.

Something like `assets/**/` or `assets/**/*`.

**Comment by fmatosqg** [comment]
Hey,

You can read the comments on the pr, there I explain why the approach you mention is a big risk for breaking backwards compatibility in the future. 

Also, web! = flutter. There are assumptions made in the specs that are unique to flutter.

**Comment by YK-Unit** [comment]
Hi, all:

I'm a member of [Fly-Mix Team](https://github.com/Fly-Mix) from NetEase company. **My team did develop a flutter development tool: `Flr(Flutter-R)` to support the feature that you want.**

`Flr` is a flutter resource manager tool, which can help flutter developer to auto specify assets in `pubspec.yaml` and generate `r.g.dart` file which likes `R.java` in Android after he changes the flutter project assets. With `r.g.dart`, flutter developer can apply the asset in code by referencing it's asset ID function, such as:

``` dart 
import 'package:flutter_r_demo/r.g.dart';

// test_sameName.png
var normalImageWidget = Image(
  width: 200,
  height: 120,
  image: R.image.test_sameName(),
);

// test_sameName.gif
var gifImageWidget = Image(
  image: R.mage.test_sameName_gif(),
);

// test.svg
var svgImageWidget = Image(
  width: 100,
  height: 100,
  image: R.svg.test(width: 100, height: 100),
);

// test.json
var jsonString = await R.text.test_json();

// test.yaml
var yamlString = await R.text.test_yaml();
```

PS: yes, it is just like the AAPT(Android Asset Packaging Tool) in Android.

**Up to now, Flr has supported Android Studio Plugin, CLI, and VSCode Extension:**

1. **Flr Android Studio Plugin** Version
    - GitHub:  https://github.com/Fly-Mix/flr-as-plugin
    
    - Plugin Homepage: https://plugins.jetbrains.com/plugin/13789-flr
    
    - flr-as-plugin Usage Example Gif: 
    
      ![flr_as_plugin-usage-example](https://user-images.githubusercontent.com/4953353/75311113-be550800-5890-11ea-85db-d01619623027.gif)

1. **Flr CLI** Version
    
    - GitHub: https://github.com/Fly-Mix/flr-cli
    
    - CLI Homepage: https://rubygems.org/gems/flr
    
    - flr-cli Usage Example Gif:
    
       ![flr_cli-usage-example](https://user-images.githubusercontent.com/4953353/75311134-d0cf4180-5890-11ea-9a2f-71d2d5d62e93.gif)

1. **Flr VSCode Extension** Version
    - GitHub: https://github.com/Fly-Mix/flr-vscode-extension
    
    - Extension Homepage: https://marketplace.visualstudio.com/items?itemName=LincolnLaw.flr
    
    - flr-vscode-extension Usage Example:
      
      ![flr_vscode_extension-usage-example](https://user-images.githubusercontent.com/4953353/75311154-dd539a00-5890-11ea-9b39-79e5f0349930.jpg)

**Comment by flutter-clutter** [comment]
Hey guys. Can we expect this to be implemented in the foreseeable future? There are numerous cases I can think of, in which an asset directory along with its sub directories and the files (or short: recursive definition of asset files) is totally necessary. It's very cumbersome to let an external tool do that.
